src/client/dfs/dfs.c

```diff
@@ -7446,3 +7442,405 @@ dfs_obj_fix_type(dfs_t *dfs, dfs_obj_t *parent, const char *name)
+	}
+
+	rc = daos_array_get_size(oh, DAOS_TX_NONE, size, NULL);
+	if (rc) {
+		daos_array_close(oh, NULL);
+		D_ERROR("daos_array_get_size() failed: "DF_RC"\n", DP_RC(rc));
+		return daos_der2errno(rc);
+	}
+
+	rc = daos_array_close(oh, NULL);
+	return daos_der2errno(rc);
+}
+
+struct dfs_pipeline {
+	daos_pipeline_t		pipeline;
+	dfs_predicate_t		pred;
+
+	mode_t			constant1;
+	mode_t			constant2;
+
+	d_iov_t			dkey_iov;
+	d_iov_t			const1_iov;
+	d_iov_t			const2_iov;
+	d_iov_t			const3_iov;
+
+	daos_filter_part_t	dkey_ft;
+	daos_filter_part_t	akey1_ft;
+	daos_filter_part_t	akey2_ft;
+	daos_filter_part_t	const0_ft;
+	daos_filter_part_t	const1_ft;
+	daos_filter_part_t	const2_ft;
+	daos_filter_part_t	const3_ft;
+	daos_filter_part_t	like_ft;
+	daos_filter_part_t	ba_ft;
+	daos_filter_part_t	eq_ft;
+	daos_filter_part_t	gt_ft;
+	daos_filter_part_t	and_ft;
+	daos_filter_part_t	or_ft;
+
+	daos_filter_t		pipef;
+};
+
+#define DKEY_F		"DAOS_FILTER_DKEY"
+#define AKEY_F		"DAOS_FILTER_AKEY"
+#define CONST_F		"DAOS_FILTER_CONST"
+#define BINARY_F	"DAOS_FILTER_TYPE_BINARY"
+#define INT8_F		"DAOS_FILTER_TYPE_UINTEGER8"
+#define INT4_F		"DAOS_FILTER_TYPE_UINTEGER4"
+#define LIKE_F		"DAOS_FILTER_FUNC_LIKE"
+#define GT_F		"DAOS_FILTER_FUNC_GT"
+#define EQ_F		"DAOS_FILTER_FUNC_EQ"
+#define BA_F		"DAOS_FILTER_FUNC_BITAND"
+#define AND_F		"DAOS_FILTER_FUNC_AND"
+#define OR_F		"DAOS_FILTER_FUNC_OR"
+#define COND_F		"DAOS_FILTER_CONDITION"
+
+int
+dfs_pipeline_create(dfs_t *dfs, dfs_predicate_t pred, uint64_t flags, dfs_pipeline_t **_dpipe)
+{
+	daos_size_t	bin_flen = strlen(BINARY_F);
+	daos_size_t	dkey_flen = strlen(DKEY_F);
+	daos_size_t	akey_flen = strlen(AKEY_F);
+	daos_size_t	const_flen = strlen(CONST_F);
+	daos_size_t	int8_flen = strlen(INT8_F);
+	daos_size_t	int4_flen = strlen(INT4_F);
+	daos_size_t	like_flen = strlen(LIKE_F);
+	daos_size_t	gt_flen = strlen(GT_F);
+	daos_size_t	eq_flen = strlen(EQ_F);
+	daos_size_t	ba_flen = strlen(BA_F);
+	daos_size_t	and_flen = strlen(AND_F);
+	daos_size_t	or_flen = strlen(OR_F);
+	daos_size_t	cond_flen = strlen(COND_F);
+	dfs_pipeline_t	*dpipe;
+	int		rc;
+
+	D_ALLOC_PTR(dpipe);
+	if (dpipe == NULL)
+		return ENOMEM;
+
+	/** copy the user predicate conditions */
+	memcpy(&dpipe->pred, &pred, sizeof(dfs_predicate_t));
+
+	daos_pipeline_init(&dpipe->pipeline);
+
+	/** build condition for entry name */
+	if (flags & DFS_FILTER_NAME) {
+		daos_size_t	name_len;
+
+		name_len = strnlen(dpipe->pred.dp_name, DFS_MAX_NAME);
+
+		d_iov_set(&dpipe->dkey_ft.part_type, DKEY_F, dkey_flen);
+		d_iov_set(&dpipe->dkey_ft.data_type, BINARY_F, bin_flen);
+		dpipe->dkey_ft.data_len = DFS_MAX_NAME;
+
+		d_iov_set(&dpipe->const0_ft.part_type, CONST_F, const_flen);
+		d_iov_set(&dpipe->const0_ft.data_type, BINARY_F, bin_flen);
+		dpipe->const0_ft.num_constants = 1;
+		dpipe->const0_ft.constant = &dpipe->dkey_iov;
+		d_iov_set(dpipe->const0_ft.constant, dpipe->pred.dp_name, name_len);
+
+		d_iov_set(&dpipe->like_ft.part_type, LIKE_F, like_flen);
```

Comment:
so does a regular expression have to use raw string type (e.g TYPE_BINARY)?

Reply:
i believe so yes. this pipeline API could use some re-work to make it user-friendly.

Reply:
Would be simple enough to add another enum entry with the same value RAW_STRING=BINARY